File: server/openai-service.ts
import OpenAI from "openai";
import { SimulatedChatbotService } from "./simulated-chatbot-service";

// the newest OpenAI model is "gpt-5" which was released August 7, 2025. do not change this unless explicitly requested by the user

export interface ChatResponse {
  response: string;
  confidence: number;
  suggestedActions?: string[];
}

let openaiInstance: OpenAI | null = null;

function getOpenAIClient(): OpenAI | null {
  if (!openaiInstance) {
    if (!process.env.OPENAI_API_KEY) {
      return null;
    }
    openaiInstance = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  }
  return openaiInstance;
}

// Response cache for common queries (expires after 5 minutes)
const responseCache = new Map<string, { response: ChatResponse; timestamp: number }>();
const CACHE_TTL = 5 * 60 * 1000; // 5 minutes

function getCachedResponse(key: string): ChatResponse | null {
  const cached = responseCache.get(key);
  if (cached && Date.now() - cached.timestamp < CACHE_TTL) {
    return cached.response;
  }
  if (cached) {
    responseCache.delete(key);
  }
  return null;
}

function setCachedResponse(key: string, response: ChatResponse): void {
  responseCache.set(key, { response, timestamp: Date.now() });
  // Limit cache size
  if (responseCache.size > 100) {
    const oldestKey = responseCache.keys().next().value;
    if (oldestKey) responseCache.delete(oldestKey);
  }
}

// Quick response patterns for EXACT greeting matches only
const QUICK_RESPONSES: Record<string, ChatResponse> = {
  "hello": { response: "Hello! I'm Dr. Tega, your AI healthcare assistant. How can I help you today?", confidence: 1.0 },
  "hi": { response: "Hi there! I'm Dr. Tega. What health questions can I assist you with?", confidence: 1.0 },
  "hey": { response: "Hey! I'm Dr. Tega, ready to assist with your health questions.", confidence: 1.0 },
  "help": { response: "I'm Dr. Tega, here to help with:\n- Symptom analysis and health assessments\n- Medication information and guidelines\n- Lab result interpretation\n- Clinical recommendations\n- Health risk evaluations\n\nWhat would you like to know?", confidence: 1.0 },
  "thank you": { response: "You're welcome! If you have any more health questions, feel free to ask. Take care!", confidence: 1.0 },
  "thanks": { response: "You're welcome! Stay healthy and don't hesitate to ask if you need more assistance.", confidence: 1.0 },
  "bye": { response: "Goodbye! Remember to take care of your health. I'm always here if you need medical guidance.", confidence: 1.0 },
  "goodbye": { response: "Goodbye! Take care of your health. I'm here whenever you need assistance.", confidence: 1.0 },
};

function normalizeQuery(query: string): string {
  return query.toLowerCase().trim().replace(/[^\w\s]/g, '').replace(/\s+/g, ' ');
}

export class OpenAIService {
  /**
   * Get AI-powered diagnosis assistance using GPT-5
   */
  static async getDiagnosisAssistance(
    symptoms: string,
    vitals?: Record<string, any>,
    medicalHistory?: string
  ): Promise<ChatResponse> {
    try {
      const openai = getOpenAIClient();
      if (!openai) {
        return SimulatedChatbotService.getDiagnosisAssistance(symptoms, vitals, medicalHistory);
      }
      const systemPrompt = `You are Dr. Tega, an expert AI healthcare assistant specializing in medical diagnosis support. You provide evidence-based clinical insights while emphasizing that you assist, not replace, healthcare professionals.

Your responses should:
1. Analyze presented symptoms in the context of vital signs and medical history
2. Suggest possible diagnoses with confidence levels (0-100%)
3. Recommend immediate actions if urgent
4. Always recommend professional medical evaluation
5. Flag any critical findings requiring immediate attention`;

      const userMessage = `
Patient Symptoms: ${symptoms}
${vitals ? `Vital Signs: ${JSON.stringify(vitals)}` : ""}
${medicalHistory ? `Medical History: ${medicalHistory}` : ""}

Please provide diagnostic suggestions with confidence scores and recommended actions.`;

      const response = await openai.chat.completions.create({
        model: "gpt-5",
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: userMessage },
        ],
        max_completion_tokens: 2048,
      });

      const content = response.choices[0].message.content || "";
      return {
        response: content,
        confidence: 0.85,
        suggestedActions: [
          "Schedule follow-up appointment",
          "Monitor vital signs",
          "Maintain medication compliance",
        ],
      };
    } catch (error: any) {
      console.error("OpenAI API Error:", error);
      
      // Fallback to simulated response
      console.log("Falling back to simulated diagnosis assistance");
      return SimulatedChatbotService.getDiagnosisAssistance(symptoms, vitals, medicalHistory);
    }
  }

  /**
   * Get medical information response from Dr. Tega
   * Optimized for speed with caching and quick responses
   */
  static async getMedicalResponse(question: string): Promise<ChatResponse> {
    const normalizedQuestion = normalizeQuery(question);
    
    // Check for EXACT quick responses first (instant) - only for simple greetings
    const quickResponse = QUICK_RESPONSES[normalizedQuestion];
    if (quickResponse && normalizedQuestion.split(' ').length <= 3) {
      return quickResponse;
    }

    // Check cache for similar queries
    const cacheKey = normalizedQuestion.slice(0, 100);
    const cached = getCachedResponse(cacheKey);
    if (cached) {
      return cached;
    }

    try {
      const openai = getOpenAIClient();
      if (!openai) {
        return SimulatedChatbotService.getMedicalResponse(question);
      }
      
      // Optimized concise prompt for faster responses
      const systemPrompt = `You are Dr. Tega, a knowledgeable AI healthcare assistant. Provide accurate, concise, evidence-based medical information. Keep responses focused and actionable. Always recommend consulting healthcare professionals for diagnosis/treatment. Flag emergencies clearly.`;

      const response = await openai.chat.completions.create({
        model: "gpt-5",
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: question },
        ],
        max_completion_tokens: 800, // Reduced for faster responses
        temperature: 0.7,
      });

      const content = response.choices[0].message.content || "";
      const result: ChatResponse = {
        response: content,
        confidence: 0.9,
      };
      
      // Cache the response
      setCachedResponse(cacheKey, result);
      
      return result;
    } catch (error: any) {
      console.error("OpenAI API Error:", error);
      
      // Fallback to simulated response
      console.log("Falling back to simulated medical response");
      return SimulatedChatbotService.getMedicalResponse(question);
    }
  }

  /**
   * Analyze lab results with AI insights
   */
  static async analyzeLabResults(labData: string): Promise<ChatResponse> {
    try {
      const openai = getOpenAIClient();
      if (!openai) {
        return SimulatedChatbotService.analyzeLabResults(labData);
      }
      const systemPrompt = `You are Dr. Tega, specialized in analyzing laboratory results. Provide insights on:
1. What the results indicate
2. Any abnormalities and their significance
3. Disease probability assessment
4. Recommended follow-up tests
5. Clinical recommendations`;

      const response = await openai.chat.completions.create({
        model: "gpt-5",
        messages: [
          { role: "system", content: systemPrompt },
          { role: "user", content: `Please analyze these lab results: ${labData}` },
        ],
        max_completion_tokens: 1500,
      });

      const content = response.choices[0].message.content || "";
      return {
        response: content,
        confidence: 0.88,
      };
    } catch (error: any) {
      console.error("OpenAI API Error:", error);
      
      // Fallback to simulated response
      console.log("Falling back to simulated lab analysis");
      return SimulatedChatbotService.analyzeLabResults(labData);
    }
  }
}

File: server_py/services/openai_service.py
from typing import Dict, Any, Optional
import os
import json
import requests

OPENAI_AVAILABLE = True

class OpenAIService:
    def __init__(self):
        self.api_key = os.getenv("OPENAI_API_KEY")
        self.api_url = "https://api.openai.com/v1/chat/completions"
        self.model = "gpt-4o-mini"
    
    def is_available(self) -> bool:
        return self.api_key is not None
    
    async def get_diagnosis_assistance(self, symptoms: str, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        if not self.is_available():
            return self._fallback_diagnosis(symptoms, patient_data)
        
        try:
            prompt = f"""As a medical AI assistant, analyze the following patient data and symptoms:

Patient Information:
- Age: {patient_data.get('age', 'Unknown')}
- Gender: {patient_data.get('gender', 'Unknown')}
- Blood Group: {patient_data.get('blood_group') or patient_data.get('bloodGroup', 'Unknown')}
- Genotype: {patient_data.get('genotype', 'Unknown')}
- Known Allergies: {patient_data.get('allergies', 'None reported')}

Vital Signs:
- Blood Pressure: {patient_data.get('bp_systolic') or patient_data.get('bloodPressureSystolic', 'Not recorded')}/{patient_data.get('bp_diastolic') or patient_data.get('bloodPressureDiastolic', 'Not recorded')} mmHg
- Heart Rate: {patient_data.get('heart_rate') or patient_data.get('heartRate', 'Not recorded')} bpm
- Temperature: {patient_data.get('temperature', 'Not recorded')}°C

Current Symptoms: {symptoms}

Please provide:
1. Possible diagnoses with confidence levels
2. Recommended tests
3. Treatment suggestions
4. Urgency assessment"""

            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {"role": "system", "content": "You are a medical AI assistant helping healthcare professionals with clinical decision support. Always recommend consulting with a qualified physician."},
                    {"role": "user", "content": prompt}
                ],
                max_tokens=1000,
                temperature=0.3
            )
            
            return {
                "analysis": response.choices[0].message.content,
                "model": "gpt-4",
                "status": "success"
            }
        except Exception as e:
            return self._fallback_diagnosis(symptoms, patient_data)
    
    async def analyze_lab_results(self, results: str, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        if not self.is_available():
            return self._fallback_lab_analysis(results)
        
        try:
            prompt = f"""Analyze the following laboratory results for a patient:

Patient Context:
- Age: {patient_data.get('age', 'Unknown')}
- Gender: {patient_data.get('gender', 'Unknown')}

Lab Results:
{results}

Please provide:
1. Interpretation of each result (normal/abnormal)
2. Clinical significance
3. Recommended follow-up tests if any
4. Suggested actions"""

            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {"role": "system", "content": "You are a clinical laboratory specialist AI helping interpret lab results."},
                    {"role": "user", "content": prompt}
                ],
                max_tokens=800,
                temperature=0.3
            )
            
            return {
                "analysis": response.choices[0].message.content,
                "model": "gpt-4",
                "status": "success"
            }
        except Exception as e:
            return self._fallback_lab_analysis(results)
    
    async def chat_with_dr_tega(self, message: str, context: Optional[str] = None) -> str:
        if not self.is_available():
            return self._fallback_chat(message)
        
        try:
            system_prompt = """You are Dr. Tega, an AI-powered medical assistant for the Digital Doctors Assistant system. 
You help patients and healthcare providers with medical questions, appointment scheduling, and health information.
Always be professional, empathetic, and remind users to consult with qualified healthcare professionals for medical decisions.
You can help with:
- General health information
- Symptom assessment (not diagnosis)
- Medication information
- Appointment scheduling guidance
- Health tips and preventive care"""

            messages = [{"role": "system", "content": system_prompt}]
            
            if context:
                messages.append({"role": "system", "content": f"Context: {context}"})
            
            messages.append({"role": "user", "content": message})
            
            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=messages,
                max_tokens=500,
                temperature=0.7
            )
            
            return response.choices[0].message.content
        except Exception as e:
            return self._fallback_chat(message)
    
    def _fallback_diagnosis(self, symptoms: str, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "analysis": "AI analysis temporarily unavailable. Based on the symptoms provided, please consult with a healthcare professional for proper diagnosis.",
            "model": "fallback",
            "status": "fallback"
        }
    
    def _fallback_lab_analysis(self, results: str) -> Dict[str, Any]:
        return {
            "analysis": "AI lab analysis temporarily unavailable. Please have these results reviewed by a qualified laboratory specialist.",
            "model": "fallback",
            "status": "fallback"
        }
    
    def _fallback_chat(self, message: str) -> str:
        responses = {
            "hello": "Hello! I'm Dr. Tega, your AI medical assistant. How can I help you today?",
            "hi": "Hi there! I'm Dr. Tega. What can I assist you with?",
            "help": "I can help you with general health questions, appointment information, and health tips. What would you like to know?",
            "appointment": "To schedule an appointment, please use the appointment booking feature in the system or contact the front desk.",
            "emergency": "If this is a medical emergency, please call emergency services immediately or go to the nearest emergency room.",
        }
        
        message_lower = message.lower()
        for key, response in responses.items():
            if key in message_lower:
                return response
        
        return "Thank you for your message. For specific medical questions, I recommend consulting with a healthcare professional. Is there anything else I can help you with?"
